common/util/trees.h

```diff
@@ -0,0 +1,856 @@
+// replace them with something from standard library.
+
+// Type used as a placeholder for unavailable feature in type traits.
+struct UnavailableFeatureTraits {
+  static inline constexpr bool available = false;
+};
+
+// Type used as a base for available feature in type traits.
+struct FeatureTraits {
+  static inline constexpr bool available = true;
+};
+
+// TreeNodeTraits implementation details:
+
+// Alias to Node::subnodes_type if it exists.
+// Intended for use with detected_or_t.
+template <typename Node,  //
+          typename Type_ = typename Node::subnodes_type>
+using TreeNodeSubnodesType = Type_;
+
+// Defined when `Node` contains `Children()` method which returns reference to
+// a STL-like container with nodes.
+// TODO(mglb): Add support for children containers storing Node pointers.
+template <typename Node,  //
+          typename ChildrenType_ = decltype(std::declval<Node>().Children()),
+          typename =
+              std::void_t<decltype(*std::declval<Node>().Children().begin()),
+                          decltype(std::declval<Node>().Children().end())>>
+struct TreeNodeChildrenTraits : FeatureTraits {
+  // Container type which should be used for storing arrays of detached
+  // child nodes. It must be move-assignable to, and move-constructible from,
+  // a value returned by `Children()` method.
+  using container_type = detected_or_t<std::remove_reference_t<ChildrenType_>,
+                                       TreeNodeSubnodesType, Node>;
+};
+
+// Defined when `Node` contains `Value()` method.
+template <typename Node,  //
+          typename Value_ = decltype(std::declval<Node>().Value())>
+struct TreeNodeValueTraits : FeatureTraits {
+  // Type of data returned by `Value()` method.
+  using type = std::remove_reference_t<Value_>;
+  // Aliases provided for convenience
+  using reference = std::add_lvalue_reference_t<type>;
+  using const_reference = std::add_lvalue_reference_t<const type>;
+};
+
+// Defined when `Node` contains `Parent()` method which dereferences to a type
+// fulfilling NodeTraits concept.
+template <typename Node,  //
+          typename Parent_ = decltype(*std::declval<Node>().Parent()),
+          typename = std::void_t<TreeNodeChildrenTraits<Parent_>>>
+struct TreeNodeParentTraits : FeatureTraits {};
+
+// BirthRank implementation details:
+
+// BirthRank implementation supportin any container
+template <class T>
+inline static size_t BirthRank(const T& node, std::input_iterator_tag) {
+  if (node.Parent() != nullptr) {
+    size_t index = 0;
+    for (const auto& child : node.Parent()->Children()) {
+      if (&node == &child) return index;
+      ++index;
+    }
+  }
+  return 0;
+}
+
+// BirthRank optimized for random access containers
+template <class T>
+inline static size_t BirthRank(const T& node, std::random_access_iterator_tag) {
+  if (node.Parent() != nullptr) {
+    return std::distance(&(node.Parent()->Children().front()), &node);
+  }
+  return 0;
```

Comment:
// a root node has no parent, but is considered rank 0

Reply:
Added in the public function (cited one is an implementation detail).